planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+    public abstract class EventStreamBase<TKey, TEvent, TEventBase> where TEvent : IEvent<TEventBase>
+    {
+        public TKey Id { get; set; }
+
+        // There's a little bit of optimization for inline
+        // projections by doing this
+
+        // Other folks want different mechanics
+        public StreamState State { get; set; }
+
+        public IList<TEvent> Events { get; } = new List<TEvent>();
+
+
+        // Data in here while it's being appended, but moves up above when
+        // it's pushed into the database
+        public IList<TEventBase> Pending { get; } = new List<TEventBase>();
+    }
+
+    public interface IEvent<TEventBase>
+    {
+        Guid Id { get; set; }
+        int Version { get; set; }
+        long Sequence { get; set; }
+        TEventBase Data { get; set; }
+        DateTimeOffset Timestamp { get; set; }
+
+        // Additional, user-defined metadata properties
+        // like user id or correlation id or whatever
+
+    }
+
+    // This might be too heavyweight
+    public interface IMultitenantEvent<TEventBase>: IEvent<TEventBase>
+    {
+        string TenantId { get; set; }
+    }
+
+The custom stream and event types could be used to define segregations between different types of event logs within a single application as part
+of the partitioning/sharding/segmenting strategy
+
+
+
+The schema generation would conform to the shape of the base event and stream wrapper much the way that document storage does in the Doc Db features.
+
+### Table/Function Generation and Stream Identity
+
+TODO -- allow more flexibility in the identity of streams, so `int`, `long`, `Guid`, `string` just like the document types. Do more code generation
+to accomodate this.
+
+TODO -- there will be more columns & even new tables to track and coordinate the asynchronous projections
+
+### Event Store Partitioning or Sharding
+
+The event sourcing in Marten has limited scalability because of its dependence today upon appending events to 
+one table. Closely related to the storage is ways to segment the asynchronous projection support to parallelize
+the work across multiple threads or active nodes.
+
+Here are some ideas:
+
+* Allow the user to define multiple event logs in the database. This way you'd have multiple stream/event tables for completely
+  different kinds of streams
+* Postgresql native partitioning by tenant id where that's useful
+* Aggregation keys -- this is tightly coupled to the projections. If there's a projection that is done by something other than tenant id or stream id, try to persist the 
+  key for however the events/streams are grouped to stream and/or event storage. Thinking about data elements like "Region" or "Country" where you might be doing
+  aggregations across streams or tenants. Maybe that's useful?
+* Archive streams that reach some kind of termination event? Move everything to an archive table off to the side? Keeps the event log table cleaner
```

Comment:
That can be done also with Postgres partitions (by detaching the partition). For sure we need tombstone events. We may also consider "summary event". The summary event would contain the snapshot of the stream and then could be used to other aggregations to reduce the need of getting all events (eg. event that the accounting year has finished).

Reply:
I like the summary event idea, especially if we try to do some kind of "map/reduce" builders for the documents. I haven't put much thought into that either way yet.